framework/calico_framework.py

```diff
@@ -0,0 +1,661 @@
+        """
+        If None, this test is not being tracked for timeout.
+
+        Otherwise, this number will be seconds-since-epoch of the last checkpoint.
+
+        For an unstarted test, this is set if/when
+        this test refuses to start while no other tests are running.
+        If time expires in this state, that means the mesos cluster was
+        unable to provide enough offers to launch this test when no other tests were running.
+
+        For a started test, this value is reset at each task update, and represents the last time
+        this Test heard from one of its tasks.
+        """
+
+        self.additional_info = None
+        """
+        Test-wide information which will be provided during the test output.
+        Useful for explaining a test-scope failure (vs. a task failure).
+        """
+
+        self.killed = False
+        """
+        Flag, when set to True, will indicate that this test is being shutdown.
+        """
+
+    def __repr__(self):
+        s_repr = "Test(%s" % self.name
+        if self.timeout != None:
+            s_repr += (", timeout=%d" % (time.time() - self.timeout))
+        s_repr += ")"
+
+        return s_repr
+
+    @property
+    def complete(self):
+        """
+        True if the test is no longer running. This means killed tests too
+        """
+        return self.killed or self.tasks_with_state(UNFINISHED_TASK_STATES) == []
+
+    def can_run_on(self, offers):
+        """
+        Returns true if the test can run on the offers passed in.
+        Offers passed in will be ignored if they are already reserved.
+        This function also reserves offers and plans tasks as it goes along.
+        It will rollback reservations if it realizes it can't finish planning all tasks.
+        """
+        offers = [offer for offer in offers if offer.reserved_by in [None, self]]
+
+        # First sort the tasks by slave
+        tasks_by_slave = {}
+        for task in [task for task in self.tasks if task.slave != None]:
+            try:
+                tasks_by_slave[task.slave].append(task)
+            except KeyError:
+                tasks_by_slave[task.slave] = [task]
+
+        # Sort those groups by most tasks to least tasks
+        tasks_grouped_by_slave = tasks_by_slave.values()
+        tasks_grouped_by_slave = sorted(tasks_grouped_by_slave, lambda x,y: cmp(len(y), len(x)))
+
+        # Then find the remaining tasks without a specified slave requirement
+        anywhere_tasks = [task for task in self.tasks if task.slave == None]
+
+        # Sort the offers by resources (this assumes each offer is from a unique slave
+        offers_by_tasks_launchable = sorted(offers, lambda x,y: cmp(y.num_tasks_launchable, x.num_tasks_launchable))
+
+        # We now have the offers sorted by largest to smallest, and the tasks-per-slave grouped from largest to smallest.
+        try:
+            for offer, tasks in izip_longest(offers_by_tasks_launchable, tasks_by_slave.values()):
```

Comment:
>Why not catch the case where there aren't enough offers early, before you even sort the offers.

How would you accomplish that without sorting the offers and then trying to place all the tasks? With the unique-slave restrictions, I don't see how its possible to tell if a test is launchable unless you actually attempt to place all its tasks.

I could run through the same loop once without actually doing the reservation, which would improve readability since there's no rollback if it realizes it can't launch the test. But at that point, I'd have to run through the same loop again to actually make the reservations.